scalahost/src/main/scala/scala/meta/internal/scalahost/v1/online/Mirror.scala

```diff
@@ -41,20 +41,32 @@ class Mirror(val global: Global)
     compilerUnits.toList.map(_.toSource)
   }
 
+  private var cachedDatabaseKey = (g.currentRun, -1, -1)
+  private def recomputeCachedDatabaseKey() = (g.currentRun, g.currentRun.size, adhocUnits.size)
+  private var cachedDatabase: Database = null
   def database: Database = {
-    var unmappedNames = ""
-    val units = compilerUnits ++ adhocUnits
-    val databases = units.map(unit => {
-      try unit.toDatabase
-      catch {
-        case ex: Exception if ex.getMessage.startsWith("Unmapped names in") =>
-          unmappedNames += (ex.getMessage + EOL)
-          Database()
+    // NOTE: We rely on the fact that compilation units change monotonously,
```

Comment:
Cool! I was thinking how to cache the database correctly since it is convenient to instantiate the mirror during global initialization.